---
layout: page
permalink: /code/
title: code
description: A list of R packages published on CRAN or on GitHub
---


* [ggmix](http://sahirbhatnagar.com/ggmix/)  
[![Travis build status](https://travis-ci.org/sahirbhatnagar/ggmix.svg?branch=master)](https://travis-ci.org/sahirbhatnagar/ggmix)   [![Coverage status](https://codecov.io/gh/sahirbhatnagar/ggmix/branch/master/graph/badge.svg)](https://codecov.io/github/sahirbhatnagar/ggmix?branch=master)   [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/ggmix)](https://cran.r-project.org/package=ggmix) ![](http://cranlogs.r-pkg.org/badges/grand-total/ggmix?color=blue) 
A General Framework for Variable Selection in Linear Mixed Models with Applications to Genetic Studies with Structured Populations. This is a penalized regression method that selects and estimates predictors in a single random effects model.

* [sail](http://sahirbhatnagar.com/sail/)  
[![Travis build status](https://travis-ci.org/sahirbhatnagar/sail.svg?branch=master)](https://travis-ci.org/sahirbhatnagar/sail)   [![Coverage status](https://codecov.io/gh/sahirbhatnagar/sail/branch/master/graph/badge.svg)](https://codecov.io/github/sahirbhatnagar/sail?branch=master)   [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/sail)](https://cran.r-project.org/package=sail) ![](http://cranlogs.r-pkg.org/badges/grand-total/sail?color=blue)  
Sparse additive interaction learning. This software package implements the `sail` method described in the accompanying paper. This is a penalized regression method that selects and estimates both linear and non-linear interactions with a single exposure variable.  


* [eclust](http://sahirbhatnagar.com/eclust/)  
[![Travis-CI Build Status](https://travis-ci.org/sahirbhatnagar/eclust.svg?branch=master)](https://travis-ci.org/sahirbhatnagar/eclust) [![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/eclust)](https://cran.r-project.org/package=eclust) ![](http://cranlogs.r-pkg.org/badges/grand-total/eclust?color=blue)  
A Statistical Software Tool for the Analysis of High-Dimensional Interactions. This software is written in the open source software environment R. It's main functionality is to cluster a high dimensional dataset (e.g. genomics, brain imaging) by leveraging the environmental exposures. See the <a href="https://doi.org/10.1101/102475">Preprint</a> for details.

* [casebase](http://sahirbhatnagar.com/casebase/)  
[![Travis-CI Build Status](https://travis-ci.org/sahirbhatnagar/casebase.svg?branch=master)](https://travis-ci.org/sahirbhatnagar/casebase) [![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/casebase)](https://cran.r-project.org/package=casebase) ![](http://cranlogs.r-pkg.org/badges/grand-total/casebase?color=blue)  
Implements the case-base sampling approach of [Hanley and Miettinen (2009)](https://doi.org/10.2202/1557-4679.1125), [Saarela and Arjas (2015)](https://doi.org/10.1111/sjos.12125), and [Saarela (2015)](https://doi.org/10.1007/s10985-015-9352-x), for fitting flexible hazard regression models to survival data with single event type or multiple competing causes via logistic and multinomial regression. From the fitted hazard function, cumulative incidence, risk functions of time, treatment and profile can be derived. This approach accommodates any log-linear hazard function of prognostic time, treatment, and covariates, and readily allows for non-proportionality. We also provide a plot method for visualizing incidence density via population time plots. This is joint work with [Maxime Turgeon](http://turgeonmaxime.github.io/), [Olli Saarela](http://individual.utoronto.ca/osaarela/) and [James Hanley](http://www.medicine.mcgill.ca/epidemiology/hanley/)

* [manhattanly](https://cran.r-project.org/web/packages/manhattanly/)  
[![Travis-CI Build Status](https://travis-ci.org/sahirbhatnagar/manhattanly.svg?branch=master)](https://travis-ci.org/sahirbhatnagar/manhattanly) [![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/manhattanly)](https://cran.r-project.org/package=manhattanly) ![](http://cranlogs.r-pkg.org/badges/grand-total/manhattanly?color=blue)  
<a href="https://www.youtube.com/watch?v=sawf9fnbsOA" target="_blank">Watch the Plotcon talk in New York city about manhattanly</a>  
Create interactive Q-Q, manhattan and volcano plots that are usable from the R console, in the 'RStudio' viewer pane, in 'R Markdown' documents, and in 'Shiny' apps. Hover the mouse pointer over a point to show details or drag a rectangle to zoom. <!--A manhattan plot is a popular graphical method for visualizing results from high-dimensional data analysis such as a (epi)genome wide association study (GWAS or EWAS), in which p-values, Z-scores, test statistics are plotted on a scatter plot against their genomic position. Manhattan plots are used for visualizing potential regions of interest in the genome that are associated with a phenotype. Interactive manhattan plots allow the inspection of specific value (e.g. rs number or gene name) by hovering the mouse over a cell, as well as zooming into a region of the genome (e.g. a chromosome) by dragging a rectangle around the relevant area. This work is based on the 'qqman' package by Stephen Turner and the 'plotly.js' engine. It produces similar manhattan and Q-Q plots as the 'manhattan' and 'qq' functions in the 'qqman' package, with the advantage of including extra annotation information and interactive web-based visualizations directly from R. Once uploaded to a 'plotly' account, 'plotly' graphs (and the data behind them) can be viewed and modified in a web browser.-->

* [acm4r](http://cran.r-project.org/web/packages/acm4r/index.html)  
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/acm4r)](https://cran.r-project.org/package=acm4r) ![](http://cranlogs.r-pkg.org/badges/grand-total/acm4r?color=blue)  
Fragment lengths or molecular weights from pairs of lanes are compared, and a number of matching bands are calculated using the Align-and-Count Method.

### Shiny Apps

* [cdiff](https://sahir.shinyapps.io/cdiff/)    
Interactive, online web application to guide preoperative decision making for patients with fulminant Clostridium difficile colitis (FCDC) being evaluated for surgery. This calculator predicts 30-day postoperative mortality for patients with FCDC based on easily attainable pre-operative parameters.   

* [Supplementary results for our paper "Assessing transmission ratio distortion in extended families: a comparison of analysis methods"](https://sahir.shinyapps.io/TRDtest/)  
