<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width initial-scale=1" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge">

  <title>Sahir Bhatnagar | Limma Moderated and Ordinary t-statistics</title>
  <meta name="description" content="A simple, whitespace theme for academics. Based on [*folio](https://github.com/bogoli/-folio) design.
">


  <!-- Include custom icon fonts -->
  <link rel="stylesheet" href="https://sahirbhatnagar.com/assets/css/fontawesome-all.min.css">
  <link rel="stylesheet" href="https://sahirbhatnagar.com/assets/css/academicons.min.css">


  <link rel="shortcut icon" href="https://sahirbhatnagar.com/assets/img/favicon.ico">
  <link rel="stylesheet" href="//code.cdn.mozilla.net/fonts/fira.css"> 
  <link rel="stylesheet" href="/assets/css/main.css">
  <link rel="canonical" href="https://sahirbhatnagar.com/limma_ttest">
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    
    <span class="site-title">
        
        <strong>Sahir</strong> Bhatnagar
    </span>
    

    <nav class="site-nav">

      <div class="trigger">
        <!-- About -->
        <a class="page-link" href="https://sahirbhatnagar.com/">about</a>

        <!-- Pages 
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/misc/">miscellaneous</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/papers/">papers</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/projects/">projects</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/code/">code</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/talks/">talks</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/teaching/">teaching</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/team/">team</a>
          
        
          
        
          
         -->

        
          
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/papers/">papers</a>
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        

        
          
        
          
        
          
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/code/">code</a>
          
        
          
        
          
        
          
        
          
        
          
        

        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/team/">team</a>
          
        
          
        
          
        


        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/teaching/">teaching</a>
          
        
          
        
          
        
          
        

        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/talks/">talks</a>
          
        
          
        
          
        
          
        
          
        
        <!-- Blog
        <a class="page-link" href="http://sahirbhatnagar.github.io/blog">blog</a>-->

        <!-- Blog--> 
        <a class="page-link" href="https://sahirbhatnagar.com/blog/">blog</a> 

        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/misc/">misc</a>
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        



        <!-- CV link -->
        <a class="page-link" href="https://sahirbhatnagar.com/assets/pdf/Bhatnagar_Research_Teaching.pdf">CV</a>
 
      </div>
    </nav>

  </div>

</header>



    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <header class="post-header">
    <h1 class="post-title">Limma Moderated and Ordinary t-statistics</h1>
    <p class="post-meta">February 7, 2017 â€¢ sahir</p>
  </header>

  <article class="post-content">
    <p>When analyzing large amounts of genetic and genomic data, the first line of analysis is usually some sort of univariate test. That is, conduct a statistical test for each SNP or CpG site or Gene and then correct for multiple testing. The <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a> package on Bioconductor is a popular method for computing <em>moderated</em> t-statistics using a combination of the <code class="language-plaintext highlighter-rouge">limma::lmFit</code> and <code class="language-plaintext highlighter-rouge">limma::eBayes</code> functions. In this post, I show how to calculate the <em>ordinary</em> t-statistics from <code class="language-plaintext highlighter-rouge">limma</code> output.</p>

<!--more-->

<p>First we load the required packages</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># clear workspace</span><span class="w">
</span><span class="n">rm</span><span class="p">(</span><span class="n">list</span><span class="o">=</span><span class="n">ls</span><span class="p">())</span><span class="w">

</span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="n">requireNamespace</span><span class="p">(</span><span class="s2">"pacman"</span><span class="p">,</span><span class="w"> </span><span class="n">quietly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span><span class="w"> 
  </span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"pacman"</span><span class="p">)</span><span class="w">

</span><span class="c1"># this will install (if you dont already have it) and load the package</span><span class="w">
</span><span class="n">pacman</span><span class="o">::</span><span class="n">p_load</span><span class="p">(</span><span class="s2">"minfi"</span><span class="p">)</span><span class="w">
</span><span class="n">pacman</span><span class="o">::</span><span class="n">p_load</span><span class="p">(</span><span class="s2">"minfiData"</span><span class="p">)</span><span class="w">
</span><span class="n">pacman</span><span class="o">::</span><span class="n">p_load</span><span class="p">(</span><span class="s2">"limma"</span><span class="p">)</span><span class="w">
</span><span class="n">pacman</span><span class="o">::</span><span class="n">p_load</span><span class="p">(</span><span class="s2">"CpGassoc"</span><span class="p">)</span><span class="w">

</span><span class="c1"># check the data that is available in the minfiData package</span><span class="w">
</span><span class="n">pacman</span><span class="o">::</span><span class="n">p_data</span><span class="p">(</span><span class="s2">"minfiData"</span><span class="p">)</span></code></pre></figure>

<figure class="highlight"><pre><code class="language-text" data-lang="text">##   Data        Description                                                                           
## 1 MsetEx      An example dataset for Illumina's Human Methylation 450k dataset, after preprocessing.
## 2 MsetEx.sub  An example dataset for Illumina's Human Methylation 450k dataset, after preprocessing.
## 3 RGsetEx     An example dataset for Illumina's Human Methylation 450k dataset.                     
## 4 RGsetEx.sub An example dataset for Illumina's Human Methylation 450k dataset.</code></pre></figure>

<p>Next, we extract some sample data and create a covariate of interest</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># get the M-values for the sample data</span><span class="w">
</span><span class="n">DT</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">minfi</span><span class="o">::</span><span class="n">getM</span><span class="p">(</span><span class="n">MsetEx.sub</span><span class="p">)</span><span class="w">

</span><span class="nf">dim</span><span class="p">(</span><span class="n">DT</span><span class="p">)</span></code></pre></figure>

<figure class="highlight"><pre><code class="language-text" data-lang="text">## [1] 600   6</code></pre></figure>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># rows are CpGs and columns are samples</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">DT</span><span class="p">)</span></code></pre></figure>

<figure class="highlight"><pre><code class="language-text" data-lang="text">##            5723646052_R02C02 5723646052_R04C01 5723646052_R05C02
## cg00050873          3.502348         0.4414491          4.340695
## cg00212031         -3.273751         0.9234662         -2.614777
## cg00213748          2.076816        -0.1309465          1.260995
## cg00214611         -3.438838         1.7463950         -2.270551
## cg00455876          1.839010        -0.9927320          1.619479
## cg01707559         -3.303987        -0.6433201         -3.540887
##            5723646053_R04C02 5723646053_R05C02 5723646053_R06C02
## cg00050873        0.24458355        -0.3219281         0.2744392
## cg00212031       -0.21052257        -0.6861413        -0.1397595
## cg00213748       -1.10373279        -1.6616553        -0.1270869
## cg00214611        0.29029649        -0.2103599        -0.6138630
## cg00455876       -0.09504721        -0.2854655         0.6361273
## cg01707559       -0.74835377        -0.4678048        -1.1345421</code></pre></figure>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># create a fake covariate. 3 intact and 3 degraded samples</span><span class="w">
</span><span class="n">tissue</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s2">"Intact"</span><span class="p">,</span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"Degraded"</span><span class="p">,</span><span class="m">3</span><span class="p">)))</span><span class="w">
</span><span class="n">design</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">model.matrix</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">tissue</span><span class="p">)</span></code></pre></figure>

<p>Then we calculate the moderated and ordinary t-statistics and compare them:</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># limma fit</span><span class="w">
</span><span class="n">fit</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">lmFit</span><span class="p">(</span><span class="n">DT</span><span class="p">,</span><span class="w"> </span><span class="n">design</span><span class="p">)</span><span class="w">
</span><span class="n">fit_ebayes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">eBayes</span><span class="p">(</span><span class="n">fit</span><span class="p">)</span><span class="w">

</span><span class="c1"># ordinary t-statistic</span><span class="w">
</span><span class="n">ordinary.t</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fit</span><span class="o">$</span><span class="n">coefficients</span><span class="p">[,</span><span class="s2">"tissueIntact"</span><span class="p">]</span><span class="w"> </span><span class="o">/</span><span class="w"> 
  </span><span class="n">fit</span><span class="o">$</span><span class="n">stdev.unscaled</span><span class="p">[,</span><span class="s2">"tissueIntact"</span><span class="p">]</span><span class="w"> </span><span class="o">/</span><span class="w"> 
  </span><span class="n">fit</span><span class="o">$</span><span class="n">sigma</span><span class="w">

</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ordinary.t</span><span class="p">,</span><span class="w"> 
     </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fit_ebayes</span><span class="o">$</span><span class="n">t</span><span class="p">[,</span><span class="s2">"tissueIntact"</span><span class="p">],</span><span class="w"> 
     </span><span class="n">xlab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ordinary t-statistic"</span><span class="p">,</span><span class="w"> 
     </span><span class="n">ylab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"moderated t-statistic"</span><span class="p">)</span><span class="w">
</span><span class="n">abline</span><span class="p">(</span><span class="n">a</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"red"</span><span class="p">)</span></code></pre></figure>

<p><img src="/figure/posts/2017-02-07-ttest_limma/unnamed-chunk-2-1.png" alt="plot of chunk unnamed-chunk-2" /></p>

<p>We can calculate the corresponding p-values from the ordinary t-statistics. This is given by</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># t-distribution with n-p-1 degrees of freedom </span><span class="w">
</span><span class="c1"># (n: number of samples, p:number of covariates excluding intercept)</span><span class="w">
</span><span class="n">ordinary.pvalue</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">pt</span><span class="p">(</span><span class="n">q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">abs</span><span class="p">(</span><span class="n">ordinary.t</span><span class="p">),</span><span class="w"> 
                      </span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">dim</span><span class="p">(</span><span class="n">DT</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="o">-</span><span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">design</span><span class="p">)</span><span class="m">-1</span><span class="p">)</span><span class="m">-1</span><span class="p">,</span><span class="w"> 
                      </span><span class="n">lower.tail</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">2</span></code></pre></figure>

<p>We can also use the <a href="https://cran.r-project.org/package=CpGassoc"><code class="language-plaintext highlighter-rouge">CpGassoc</code></a> package to calculate ordinary t-statistics and compare the result to our manual calculations:</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># compare result with CpGassoc package (regular t-test)</span><span class="w">
</span><span class="n">t_test_mvalues</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">CpGassoc</span><span class="o">::</span><span class="n">cpg.assoc</span><span class="p">(</span><span class="w">
  </span><span class="n">beta.val</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">minfi</span><span class="o">::</span><span class="n">getBeta</span><span class="p">(</span><span class="n">MsetEx.sub</span><span class="p">),</span><span class="w">
  </span><span class="n">indep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">tissue</span><span class="p">,</span><span class="w">
  </span><span class="n">fdr.cutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w">
  </span><span class="n">logit.transform</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="c1"># these should be equal</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">t_test_mvalues</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">P.value</span><span class="p">,</span><span class="w"> </span><span class="n">ordinary.pvalue</span><span class="p">)</span><span class="w">
</span><span class="n">abline</span><span class="p">(</span><span class="n">a</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"red"</span><span class="p">)</span></code></pre></figure>

<p><img src="/figure/posts/2017-02-07-ttest_limma/unnamed-chunk-4-1.png" alt="plot of chunk unnamed-chunk-4" /></p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># t-statistic is the squared F.statistic (in certain cases) </span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">t_test_mvalues</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">F.statistic</span><span class="p">,</span><span class="w"> </span><span class="n">ordinary.t</span><span class="o">^</span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="n">abline</span><span class="p">(</span><span class="n">a</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"red"</span><span class="p">)</span></code></pre></figure>

<p><img src="/figure/posts/2017-02-07-ttest_limma/unnamed-chunk-4-2.png" alt="plot of chunk unnamed-chunk-4" /></p>

<h2 id="session-info">Session Info</h2>

<figure class="highlight"><pre><code class="language-text" data-lang="text">## R version 3.3.2 (2016-10-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.1 LTS
## 
## attached base packages:
## [1] stats4    parallel  methods   stats     graphics  grDevices
## [7] utils     datasets  base     
## 
## other attached packages:
##  [1] CpGassoc_2.55                                     
##  [2] nlme_3.1-128                                      
##  [3] limma_3.30.10                                     
##  [4] minfiData_0.20.0                                  
##  [5] IlluminaHumanMethylation450kanno.ilmn12.hg19_0.6.0
##  [6] IlluminaHumanMethylation450kmanifest_0.4.0        
##  [7] minfi_1.20.2                                      
##  [8] bumphunter_1.14.0                                 
##  [9] locfit_1.5-9.1                                    
## [10] iterators_1.0.8                                   
## [11] foreach_1.4.3                                     
## [12] Biostrings_2.42.1                                 
## [13] XVector_0.14.0                                    
## [14] SummarizedExperiment_1.4.0                        
## [15] GenomicRanges_1.26.2                              
## [16] GenomeInfoDb_1.10.2                               
## [17] IRanges_2.8.1                                     
## [18] S4Vectors_0.12.1                                  
## [19] Biobase_2.34.0                                    
## [20] BiocGenerics_0.20.0                               
## [21] knitr_1.15.1                                      
## 
## loaded via a namespace (and not attached):
##  [1] mclust_5.2.2             base64_2.0              
##  [3] Rcpp_0.12.9              lattice_0.20-34         
##  [5] Rsamtools_1.26.1         digest_0.6.12           
##  [7] R6_2.2.0                 plyr_1.8.4              
##  [9] RSQLite_1.1-2            evaluate_0.10           
## [11] highr_0.6                httr_1.2.1              
## [13] zlibbioc_1.20.0          GenomicFeatures_1.26.2  
## [15] data.table_1.10.4        annotate_1.52.1         
## [17] Matrix_1.2-7.1           preprocessCore_1.36.0   
## [19] splines_3.3.2            BiocParallel_1.8.1      
## [21] stringr_1.2.0            RCurl_1.95-4.8          
## [23] biomaRt_2.30.0           rtracklayer_1.34.1      
## [25] multtest_2.30.0          pkgmaker_0.22           
## [27] openssl_0.9.6            GEOquery_2.40.0         
## [29] quadprog_1.5-5           codetools_0.2-15        
## [31] matrixStats_0.51.0       XML_3.98-1.5            
## [33] reshape_0.8.6            GenomicAlignments_1.10.0
## [35] MASS_7.3-45              bitops_1.0-6            
## [37] grid_3.3.2               xtable_1.8-2            
## [39] registry_0.3             DBI_0.5-1               
## [41] pacman_0.4.1             magrittr_1.5            
## [43] stringi_1.1.2            genefilter_1.56.0       
## [45] doRNG_1.6                nor1mix_1.2-2           
## [47] RColorBrewer_1.1-2       siggenes_1.48.0         
## [49] tools_3.3.2              illuminaio_0.16.0       
## [51] rngtools_1.2.4           survival_2.40-1         
## [53] AnnotationDbi_1.36.2     beanplot_1.2            
## [55] memoise_1.0.0</code></pre></figure>


  </article>

  
    <div id="disqus_thread"></div>
    <script type="text/javascript">
      var disqus_shortname  = 'al-folio';
      var disqus_identifier = '/ttest_limma';
      var disqus_title      = "Limma Moderated and Ordinary t-statistics";
      (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
      })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
  

</div>

      </div>
    </div>

    <footer>

  <div class="wrapper">
    &copy; Copyright 2022 Sahir Bhatnagar.
    Powered by <a href="http://jekyllrb.com/" target="_blank">Jekyll</a>, <a href="https://github.com/alshedivat/al-folio">al-folio</a>, <a href="http://daringfireball.net/projects/markdown/">markdown</a>, <a href="https://cran.r-project.org/">R</a>, <a href="http://yihui.name/knitr/">knitr</a>, <a href="http://johnmacfarlane.net/pandoc/">pandoc</a>, <a href="https://github.com/mozilla/Fira">Fira</a> and <a href="https://github.com/sahirbhatnagar/sahirbhatnagar.github.io">GitHub</a>

    
  </div>

</footer>


    <!-- Load jQuery -->
<script src="//code.jquery.com/jquery-1.12.4.min.js"></script>

<!-- Load Common JS -->
<script src="https://sahirbhatnagar.com/assets/js/common.js"></script>


<!-- Load KaTeX -->
<link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/KaTeX/0.7.1/katex.min.css">
<script src="//cdnjs.cloudflare.com/ajax/libs/KaTeX/0.7.1/katex.min.js"></script>
<script src="https://sahirbhatnagar.com/assets/js/katex.js"></script>




<!-- Include custom icon fonts -->
<link rel="stylesheet" href="https://sahirbhatnagar.com/assets/css/font-awesome.min.css">
<link rel="stylesheet" href="https://sahirbhatnagar.com/assets/css/academicons.min.css">

<link href="//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css" rel="stylesheet">
<link rel="stylesheet" href="https://cdn.rawgit.com/jpswalsh/academicons/master/css/academicons.min.css">


<!-- Google Analytics -->
<script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
})(window,document,'script','//www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-60189402-1', 'auto');
ga('send', 'pageview');
</script>


  </body>

</html>
