<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width initial-scale=1" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge">

  <title>Sahir Bhatnagar | Mixed Models with Kinship in R</title>
  <meta name="description" content="A simple, whitespace theme for academics. Based on [*folio](https://github.com/bogoli/-folio) design.
">

  <link rel="shortcut icon" href="https://sahirbhatnagar.com/assets/img/favicon.ico">
  <link rel="stylesheet" href="//code.cdn.mozilla.net/fonts/fira.css"> 
  <link rel="stylesheet" href="/assets/css/main.css">
  <link rel="canonical" href="https://sahirbhatnagar.com/blog/2017/lmm_kinship_r/">
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    
    <span class="site-title">
        
        <strong>Sahir</strong> Bhatnagar
    </span>
    

    <nav class="site-nav">

      <div class="trigger">
        <!-- About -->
        <a class="page-link" href="https://sahirbhatnagar.com/">about</a>

        <!-- Pages 
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/misc/">miscellaneous</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/papers/">papers</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/projects/">projects</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/code/">code</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/talks/">talks</a>
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/teaching/">teaching</a>
          
        
          
        
          
         -->

        
          
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/papers/">papers</a>
          
        
          
        
          
        
          
        
          
        
          
        
          
        

        
          
        
          
        
          
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/code/">code</a>
          
        
          
        
          
        
          
        
          
        

        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/teaching/">teaching</a>
          
        
          
        
          
        

        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/talks/">talks</a>
          
        
          
        
          
        
          
        
        <!-- Blog
        <a class="page-link" href="http://sahirbhatnagar.github.io/blog">blog</a>-->

        <!-- Blog--> 
        <a class="page-link" href="https://sahirbhatnagar.com/blog/">blog</a> 

        
          
        
          
        
          
        
          
            <a class="page-link" href="https://sahirbhatnagar.com/misc/">misc</a>
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        



        <!-- CV link -->
        <a class="page-link" href="https://sahirbhatnagar.com/assets/pdf/BhatnagarCV2018.pdf">CV</a>

      </div>
    </nav>

  </div>

</header>



    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <header class="post-header">
    <h1 class="post-title">Mixed Models with Kinship in R</h1>
    <p class="post-meta">October 12, 2017</p>
  </header>

  <article class="post-content">
    <p>In this post, I describe how to estimate a kinship matrix and subsequently fit a mixed model using that estimated kinship. In particular, I show how this can be done on an arbitrary matrix of genotype data, which is not stored in <code class="highlighter-rouge">plink</code> format. I also show how to deal with missing genotypes.</p>

<h2 id="load-packages">Load Packages</h2>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># install.packages("gaston")</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">gaston</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h2 id="set-simulation-parameters">Set Simulation Parameters</h2>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># number of subjects</span><span class="w">
</span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1000</span><span class="w">

</span><span class="c1"># number of genotypes</span><span class="w">
</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1e4</span><span class="w">

</span><span class="c1"># number of causal genotypes</span><span class="w">
</span><span class="n">p_causal</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">50</span><span class="w">

</span><span class="c1"># Signal to noise ratio</span><span class="w">
</span><span class="n">signal_to_noise_ratio</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">2</span><span class="w">

</span><span class="c1"># vector of allele frequencies from which to sample</span><span class="w">
</span><span class="n">probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="m">0.1</span><span class="p">,</span><span class="w"> </span><span class="m">0.3</span><span class="p">,</span><span class="w"> </span><span class="m">0.4</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h2 id="generate-sample-data-with-missing-genotypes">Generate Sample Data with Missing Genotypes</h2>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">345321</span><span class="p">)</span><span class="w">
</span><span class="n">geno</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">replicate</span><span class="p">(</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">rbinom</span><span class="p">(</span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">sample</span><span class="p">(</span><span class="n">probs</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)))</span><span class="w">
</span><span class="nf">dim</span><span class="p">(</span><span class="n">geno</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1]  1000 10000
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">geno</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##      [,1] [,2] [,3] [,4] [,5]
## [1,]    0    0    1    1    0
## [2,]    1    0    0    1    0
## [3,]    1    1    0    0    0
## [4,]    2    1    1    0    0
## [5,]    1    2    1    0    0
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">geno</span><span class="p">[</span><span class="n">sample</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">NA</span><span class="w">
</span><span class="n">geno</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##      [,1] [,2] [,3] [,4] [,5]
## [1,]    0    0    1    1    0
## [2,]    1    0    0    1    0
## [3,]    1    1    0    0    0
## [4,]    2    1    1    0    0
## [5,]    1    2    1    0    0
</code></pre></div></div>

<h2 id="convert-to-bed-matrix">Convert to BED Matrix</h2>

<p>This is so that we can use the functions in the <code class="highlighter-rouge">gaston</code> package for fitting mixed models</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">DT</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gaston</span><span class="o">::</span><span class="n">as.bed.matrix</span><span class="p">(</span><span class="n">geno</span><span class="p">)</span><span class="w">

</span><span class="c1"># can access the data by </span><span class="w">
</span><span class="n">as.matrix</span><span class="p">(</span><span class="n">DT</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##   M_1 M_2 M_3 M_4 M_5
## 1   0   0   1   1   0
## 2   1   0   0   1   0
## 3   1   1   0   0   0
## 4   2   1   1   0   0
## 5   1   2   1   0   0
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># see the contents of DT</span><span class="w">
</span><span class="n">slotNames</span><span class="p">(</span><span class="n">DT</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] "ped"                  "snps"                 "bed"                 
## [4] "p"                    "mu"                   "sigma"               
## [7] "standardize_p"        "standardize_mu_sigma"
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># to access the different contents of DT use @ </span><span class="w">
</span><span class="n">DT</span><span class="o">@</span><span class="n">snps</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##   chr  id dist pos A1 A2  N0  N1  N2 NAs N0.f N1.f N2.f NAs.f callrate
## 1  NA M_1   NA  NA NA NA 480 411 101   8   NA   NA   NA    NA    0.992
## 2  NA M_2   NA  NA NA NA 349 477 167   7   NA   NA   NA    NA    0.993
## 3  NA M_3   NA  NA NA NA 461 442  85  12   NA   NA   NA    NA    0.988
## 4  NA M_4   NA  NA NA NA 781 200   9  10   NA   NA   NA    NA    0.990
## 5  NA M_5   NA  NA NA NA 886  95   3  16   NA   NA   NA    NA    0.984
##          maf         hz
## 1 0.30897177 0.41431452
## 2 0.40835851 0.48036254
## 3 0.30971660 0.44736842
## 4 0.11010101 0.20202020
## 5 0.05132114 0.09654472
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">DT</span><span class="o">@</span><span class="n">ped</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##   famid id father mother sex pheno   N0   N1  N2 NAs N0.x N1.x N2.x NAs.x
## 1     1  1      0      0   0     0 6431 2916 652   1    0    0    0     0
## 2     2  2      0      0   0     0 6450 2873 677   0    0    0    0     0
## 3     3  3      0      0   0     0 6441 2899 660   0    0    0    0     0
## 4     4  4      0      0   0     0 6374 2986 640   0    0    0    0     0
## 5     5  5      0      0   0     0 6339 2978 683   0    0    0    0     0
##   N0.y N1.y N2.y NAs.y N0.mt N1.mt N2.mt NAs.mt callrate    hz callrate.x
## 1    0    0    0     0     0     0     0      0     -Inf -2916        NaN
## 2    0    0    0     0     0     0     0      0      NaN   Inf        NaN
## 3    0    0    0     0     0     0     0      0      NaN   Inf        NaN
## 4    0    0    0     0     0     0     0      0      NaN   Inf        NaN
## 5    0    0    0     0     0     0     0      0      NaN   Inf        NaN
##   hz.x callrate.y hz.y callrate.mt hz.mt
## 1  NaN        NaN  NaN         NaN   NaN
## 2  NaN        NaN  NaN         NaN   NaN
## 3  NaN        NaN  NaN         NaN   NaN
## 4  NaN        NaN  NaN         NaN   NaN
## 5  NaN        NaN  NaN         NaN   NaN
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># p contains the alternate allele frequency </span><span class="w">
</span><span class="c1"># mu is equal to 2*p and is the expected value of the genotype (coded in 0, 1, 2)</span><span class="w">
</span><span class="c1"># sigma is the genotype standard error</span><span class="w">
</span><span class="n">DT</span><span class="o">@</span><span class="n">p</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##  [1] 0.30897177 0.40835851 0.30971660 0.11010101 0.05132114 0.05393145
##  [7] 0.28556911 0.04984894 0.30502513 0.39635996
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">DT</span><span class="o">@</span><span class="n">mu</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##  [1] 0.61794355 0.81671702 0.61943320 0.22020202 0.10264228 0.10786290
##  [7] 0.57113821 0.09969789 0.61005025 0.79271992
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">DT</span><span class="o">@</span><span class="n">sigma</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##  [1] 0.6607853 0.6950724 0.6350671 0.4338020 0.3110142 0.3155285 0.6256291
##  [8] 0.3053246 0.6458457 0.6949752
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="m">2</span><span class="o">*</span><span class="n">DT</span><span class="o">@</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">DT</span><span class="o">@</span><span class="n">mu</span><span class="p">)</span><span class="w">
</span><span class="n">abline</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="m">0</span><span class="p">,</span><span class="n">b</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"red"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="/figure/posts/lmm_example_files/figure-html/unnamed-chunk-4-1.png" alt="" /><!-- --></p>

<p>If the Hardy-Weinberg equilibrium holds, <code class="highlighter-rouge">sigma</code> should be close to <script type="math/tex">\sqrt{2*p(1-p)}</script>. This is illustrated on the figure below</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">DT</span><span class="o">@</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">DT</span><span class="o">@</span><span class="n">sigma</span><span class="p">,</span><span class="w"> </span><span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">))</span><span class="w">
</span><span class="n">t</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">seq</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="n">length</span><span class="o">=</span><span class="m">101</span><span class="p">);</span><span class="w">
</span><span class="n">lines</span><span class="p">(</span><span class="n">t</span><span class="p">,</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="m">2</span><span class="o">*</span><span class="n">t</span><span class="o">*</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">t</span><span class="p">)),</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s2">"red"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="/figure/posts/lmm_example_files/figure-html/unnamed-chunk-5-1.png" alt="" /><!-- --></p>

<h2 id="standardized-snp-matrix">Standardized SNP matrix</h2>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># this will center the columns of DT to mean 0, and standard deviation sqrt(2p(1-p))</span><span class="w">
</span><span class="n">gaston</span><span class="o">::</span><span class="n">standardize</span><span class="p">(</span><span class="n">DT</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"p"</span><span class="w">
</span><span class="n">X</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">DT</span><span class="p">)</span><span class="w">
</span><span class="n">X</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##          M_1        M_2        M_3        M_4      M_5
## 1 -0.9456415 -1.1749151  0.5819959  1.7615748 -0.32893
## 2  0.5846625 -1.1749151 -0.9472912  1.7615748 -0.32893
## 3  0.5846625  0.2636678 -0.9472912 -0.4974395 -0.32893
## 4  2.1149665  0.2636678  0.5819959 -0.4974395 -0.32893
## 5  0.5846625  1.7022506  0.5819959 -0.4974395 -0.32893
</code></pre></div></div>

<h2 id="dealing-with-missing-values">Dealing With Missing Values</h2>

<p>In standardized matrices, the <code class="highlighter-rouge">NA</code> values are replaced by zeroes, which amount to impute the missing genotypes by the mean genotype.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">X</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">X</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">
</span><span class="n">X</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##          M_1        M_2        M_3        M_4      M_5
## 1 -0.9456415 -1.1749151  0.5819959  1.7615748 -0.32893
## 2  0.5846625 -1.1749151 -0.9472912  1.7615748 -0.32893
## 3  0.5846625  0.2636678 -0.9472912 -0.4974395 -0.32893
## 4  2.1149665  0.2636678  0.5819959 -0.4974395 -0.32893
## 5  0.5846625  1.7022506  0.5819959 -0.4974395 -0.32893
</code></pre></div></div>

<p>The object <code class="highlighter-rouge">X</code> is what will be used as the data matrix in the LMM analysis. We also need to create the kinship matrix, which we do next.</p>

<h2 id="calculate-kinship-matrix">Calculate Kinship Matrix</h2>

<p>If <script type="math/tex">X_s</script> is a standardized <script type="math/tex">n \times p</script> matrix of genotypes, a Genetic Relationship Matrix of individuals can be computed as</p>

<script type="math/tex; mode=display">GRM = \frac{1}{p-1} X_s X_s^\top</script>

<p>where <script type="math/tex">p</script> is the number of SNPs and <script type="math/tex">n</script> is the number of individuals. This computation is done by the <code class="highlighter-rouge">gaston::GRM</code> function. Note that we could also use</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="p">(</span><span class="m">1</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="n">p</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="m">1</span><span class="p">))</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">tcrossprod</span><span class="p">(</span><span class="n">X</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>to calculate the kinship (covariance) matrix, but the <code class="highlighter-rouge">gaston::GRM</code> function is faster.</p>

<p>Note that the <code class="highlighter-rouge">gaston::GRM</code> function internally standardizes the genotype data, which is why we provide it the object <code class="highlighter-rouge">DT</code>. The object <code class="highlighter-rouge">X</code> will be used for fitting the model. We specify <code class="highlighter-rouge">autosome.only = FALSE</code> because we don’t have that information.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">kin</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gaston</span><span class="o">::</span><span class="n">GRM</span><span class="p">(</span><span class="n">DT</span><span class="p">,</span><span class="w"> </span><span class="n">autosome.only</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">kin</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##              1            2             3             4             5
## 1  0.994037390 -0.020637164 -2.610047e-03  1.937922e-02 -0.0050506615
## 2 -0.020637164  0.970255971  1.689422e-03 -1.211568e-02  0.0014872842
## 3 -0.002610047  0.001689422  9.864578e-01 -9.975582e-05  0.0006858376
## 4  0.019379225 -0.012115683 -9.975582e-05  9.979130e-01  0.0061614369
## 5 -0.005050662  0.001487284  6.858376e-04  6.161437e-03  1.0134803057
</code></pre></div></div>

<h2 id="principal-components">Principal Components</h2>

<p>From the GRM, we can compute the Principal components. The eigenvectors are normalized. The Principal Components (PC) can be computed by multiplying them by the square root of the associated eigenvalues</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">eiK</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">eigen</span><span class="p">(</span><span class="n">kin</span><span class="p">)</span><span class="w">

</span><span class="c1"># deal with a small negative eigen value</span><span class="w">
</span><span class="n">eiK</span><span class="o">$</span><span class="n">values</span><span class="p">[</span><span class="w"> </span><span class="n">eiK</span><span class="o">$</span><span class="n">values</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">

</span><span class="n">PC</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sweep</span><span class="p">(</span><span class="n">eiK</span><span class="o">$</span><span class="n">vectors</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="n">eiK</span><span class="o">$</span><span class="n">values</span><span class="p">),</span><span class="w"> </span><span class="s2">"*"</span><span class="p">)</span><span class="w">
</span><span class="nf">dim</span><span class="p">(</span><span class="n">PC</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] 1000 1000
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">PC</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">PC</span><span class="p">[,</span><span class="m">2</span><span class="p">])</span><span class="w">
</span></code></pre></div></div>

<p><img src="/figure/posts/lmm_example_files/figure-html/unnamed-chunk-10-1.png" alt="" /><!-- --></p>

<h2 id="simulate-phenotype">Simulate Phenotype</h2>

<p><code class="highlighter-rouge">p_causal</code> SNPs are randomly assigned to a Uniform(0.9,1.1) distribution.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">beta</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="p">)</span><span class="w">
</span><span class="n">beta</span><span class="p">[</span><span class="n">sample</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">p_causal</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">runif</span><span class="p">(</span><span class="n">p_causal</span><span class="p">,</span><span class="w"> </span><span class="n">min</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.9</span><span class="p">,</span><span class="w"> </span><span class="n">max</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.1</span><span class="p">)</span><span class="w">

</span><span class="n">y.star</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">X</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="n">beta</span><span class="w"> 

</span><span class="n">error</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">stats</span><span class="o">::</span><span class="n">rnorm</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="w">

</span><span class="n">k</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">sqrt</span><span class="p">(</span><span class="n">stats</span><span class="o">::</span><span class="n">var</span><span class="p">(</span><span class="n">y.star</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">signal_to_noise_ratio</span><span class="o">*</span><span class="n">stats</span><span class="o">::</span><span class="n">var</span><span class="p">(</span><span class="n">error</span><span class="p">))))</span><span class="w">

</span><span class="n">Y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">y.star</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">k</span><span class="o">*</span><span class="n">error</span><span class="w">
</span></code></pre></div></div>

<h2 id="run-univariate-lmm">Run Univariate LMM</h2>

<p>There are two packages we can use to fit uni/multivariate LMMs.</p>

<h3 id="gaston-package"><code class="highlighter-rouge">gaston</code> package</h3>

<p>Here we use the <code class="highlighter-rouge">gaston</code> package <a href="#gaston">(Perdry &amp; Dandine-Roulland, 2017)</a>.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># make design matrix with intercept</span><span class="w">
</span><span class="n">x</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cbind</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">X</span><span class="p">[,</span><span class="m">1</span><span class="p">,</span><span class="n">drop</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">])</span><span class="w">

</span><span class="c1"># with 1 random effect this function is faster than lmm.aireml</span><span class="w">
</span><span class="c1"># in gaston::lmm.diago you provide the eigen decomposition </span><span class="w">
</span><span class="c1"># in gaston::lmm.aireml you provide the kinship matrix</span><span class="w">
</span><span class="n">fit</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gaston</span><span class="o">::</span><span class="n">lmm.diago</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">eigenK</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eiK</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## Optimization in interval [0, 1]
## Optimizing with p = 0
## [Iteration 1] Current point = 0 df = 28.579
## [Iteration 2] Current point = 0.54494 df = 2.00363
## [Iteration 3] Current point = 0.585132 df = -0.018836
## [Iteration 4] Current point = 0.584762 df = -1.81023e-06
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># equivalently you can also fit using </span><span class="w">
</span><span class="c1"># fit &lt;- gaston::lmm.aireml(Y, x1, K = kin, verbose = FALSE)</span><span class="w">

</span><span class="c1"># the second coefficient is x1, the first is the intercept</span><span class="w">
</span><span class="p">(</span><span class="n">z_score</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fit</span><span class="o">$</span><span class="n">BLUP_beta</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="o">/</span><span class="nf">sqrt</span><span class="p">(</span><span class="n">fit</span><span class="o">$</span><span class="n">varbeta</span><span class="p">[</span><span class="m">2</span><span class="p">,</span><span class="m">2</span><span class="p">]))</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] 0.4642883
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># pvalue</span><span class="w">
</span><span class="m">2</span><span class="o">*</span><span class="n">pnorm</span><span class="p">(</span><span class="n">z_score</span><span class="p">,</span><span class="w"> </span><span class="n">lower.tail</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] 0.6424412
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># random effect variance</span><span class="w">
</span><span class="n">fit</span><span class="o">$</span><span class="n">tau</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] 44.29653
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># error variance</span><span class="w">
</span><span class="n">fit</span><span class="o">$</span><span class="n">sigma2</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] 31.45491
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># error sd</span><span class="w">
</span><span class="nf">sqrt</span><span class="p">(</span><span class="n">fit</span><span class="o">$</span><span class="n">sigma2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] 5.608468
</code></pre></div></div>

<h3 id="coxme-package"><code class="highlighter-rouge">coxme</code> package</h3>

<p>Here we use the <code class="highlighter-rouge">coxme</code> package <a href="#coxme">(Therneau, 2015)</a>.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># install.packages("coxme")</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">coxme</span><span class="p">)</span><span class="w">

</span><span class="c1"># need an ID variable</span><span class="w">
</span><span class="n">dat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="o">=</span><span class="n">X</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n</span><span class="p">)</span><span class="w">

</span><span class="c1"># provide the kinship matrix </span><span class="w">
</span><span class="n">gfit1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">lmekin</span><span class="p">(</span><span class="n">Y</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="p">(</span><span class="m">1</span><span class="o">|</span><span class="n">id</span><span class="p">),</span><span class="w"> </span><span class="n">data</span><span class="o">=</span><span class="n">dat</span><span class="p">,</span><span class="w"> </span><span class="n">varlist</span><span class="o">=</span><span class="n">kin</span><span class="p">)</span><span class="w">
</span><span class="n">gfit1</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## Linear mixed-effects kinship model fit by maximum likelihood
##   Data: dat 
##   Log-likelihood = -3572.846 
##   n= 1000 
## 
## 
## Model:  Y ~ x + (1 | id) 
## Fixed coefficients
##                 Value Std Error    z     p
## (Intercept) 0.3185476 0.1720343 1.85 0.064
## x           0.1320349 0.2748453 0.48 0.630
## 
## Random effects
##  Group Variable Std Dev   Variance 
##  id    Vmat.1    6.790932 46.116756
## Residual error= 5.440202
</code></pre></div></div>

<h2 id="references">References</h2>

<ol class="bibliography"><li>

<div id="gaston">
  
    <span class="title">gaston: Genetic Data Handling (QC, GRM, LD, PCA) &amp; Linear Mixed Models. R package version 1.5</span>
    <span class="author">
      
        
          
            
              Perdry, Hervé,
            
          
        
      
        
          and
          
            
              Dandine-Roulland, Claire
            
          
        
      
    </span>

    <span class="periodical">
    
      <em></em>
    
    
      2017
    
    </span>
  

  <span class="links">
  
  
  
  
  
  
    [<a href="https://CRAN.R-project.org/package=gaston" target="_blank">Website</a>]
  
  
  
  
  
  </span>

  <!-- Hidden abstract block -->
  
</div>
</li>
<li>

<div id="coxme">
  
    <span class="title">coxme: Mixed Effects Cox Models. R package version 2.2-5</span>
    <span class="author">
      
        
          and
          
            
              Therneau, Terry M.
            
          
        
      
    </span>

    <span class="periodical">
    
      <em></em>
    
    
      2015
    
    </span>
  

  <span class="links">
  
  
  
  
  
  
    [<a href="https://CRAN.R-project.org/package=coxme" target="_blank">Website</a>]
  
  
  
  
  
  </span>

  <!-- Hidden abstract block -->
  
</div>
</li></ol>

<h2 id="session-info">Session Info</h2>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##  setting  value                       
##  version  R version 3.4.1 (2017-06-30)
##  system   x86_64, linux-gnu           
##  ui       X11                         
##  language en_US                       
##  collate  en_US.UTF-8                 
##  tz       Canada/Eastern              
##  date     2017-10-12                  
## 
##  package      * version date       source                          
##  backports      1.1.0   2017-05-22 cran (@1.1.0)                   
##  base         * 3.4.1   2017-07-08 local                           
##  bdsmatrix    * 1.3-2   2014-08-22 CRAN (R 3.4.1)                  
##  compiler       3.4.1   2017-07-08 local                           
##  coxme        * 2.2-5   2015-06-15 CRAN (R 3.4.1)                  
##  datasets     * 3.4.1   2017-07-08 local                           
##  devtools       1.13.3  2017-08-02 CRAN (R 3.4.1)                  
##  digest         0.6.12  2017-01-27 CRAN (R 3.4.1)                  
##  evaluate       0.10.1  2017-06-24 cran (@0.10.1)                  
##  gaston       * 1.5     2017-05-25 CRAN (R 3.4.1)                  
##  graphics     * 3.4.1   2017-07-08 local                           
##  grDevices    * 3.4.1   2017-07-08 local                           
##  grid           3.4.1   2017-07-08 local                           
##  htmltools      0.3.6   2017-04-28 cran (@0.3.6)                   
##  knitr          1.17    2017-08-10 cran (@1.17)                    
##  lattice        0.20-35 2017-03-25 CRAN (R 3.3.3)                  
##  magrittr       1.5     2014-11-22 CRAN (R 3.4.1)                  
##  Matrix         1.2-11  2017-08-16 CRAN (R 3.4.1)                  
##  memoise        1.1.0   2017-04-21 CRAN (R 3.4.1)                  
##  methods      * 3.4.1   2017-07-08 local                           
##  nlme           3.1-131 2017-02-06 CRAN (R 3.4.0)                  
##  Rcpp         * 0.12.12 2017-07-15 CRAN (R 3.4.1)                  
##  RcppParallel * 4.3.20  2016-08-16 CRAN (R 3.4.1)                  
##  rmarkdown      1.6     2017-06-15 CRAN (R 3.4.1)                  
##  rprojroot      1.2     2017-01-16 CRAN (R 3.4.1)                  
##  splines        3.4.1   2017-07-08 local                           
##  stats        * 3.4.1   2017-07-08 local                           
##  stringi        1.1.5   2017-04-07 CRAN (R 3.4.1)                  
##  stringr        1.2.0   2017-02-18 CRAN (R 3.4.1)                  
##  survival     * 2.41-3  2017-04-04 CRAN (R 3.4.0)                  
##  tools          3.4.1   2017-07-08 local                           
##  utils        * 3.4.1   2017-07-08 local                           
##  withr          2.0.0   2017-09-18 Github (jimhester/withr@d1f0957)
##  yaml           2.1.14  2016-11-12 cran (@2.1.14)
</code></pre></div></div>


  </article>

  

</div>

      </div>
    </div>

    <footer>

  <div class="wrapper">
    &copy; Copyright 2018 Sahir Bhatnagar.
    Powered by <a href="http://jekyllrb.com/" target="_blank">Jekyll</a>, <a href="https://github.com/alshedivat/al-folio">al-folio</a>, <a href="http://daringfireball.net/projects/markdown/">markdown</a>, <a href="https://cran.r-project.org/">R</a>, <a href="http://yihui.name/knitr/">knitr</a>, <a href="http://johnmacfarlane.net/pandoc/">pandoc</a>, <a href="https://github.com/mozilla/Fira">Fira</a> and <a href="https://github.com/sahirbhatnagar/sahirbhatnagar.github.io">GitHub</a>

    
  </div>

</footer>


    <!-- Load jQuery -->
<script src="//code.jquery.com/jquery-1.12.4.min.js"></script>

<!-- Load Common JS -->
<script src="https://sahirbhatnagar.com/assets/js/common.js"></script>


<!-- Load KaTeX -->
<link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/KaTeX/0.7.1/katex.min.css">
<script src="//cdnjs.cloudflare.com/ajax/libs/KaTeX/0.7.1/katex.min.js"></script>
<script src="https://sahirbhatnagar.com/assets/js/katex.js"></script>




<!-- Include custom icon fonts -->
<link rel="stylesheet" href="https://sahirbhatnagar.com/assets/css/font-awesome.min.css">
<link rel="stylesheet" href="https://sahirbhatnagar.com/assets/css/academicons.min.css">

<link href="//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css" rel="stylesheet">
<link rel="stylesheet" href="https://cdn.rawgit.com/jpswalsh/academicons/master/css/academicons.min.css">


<!-- Google Analytics -->
<script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
})(window,document,'script','//www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-60189402-1', 'auto');
ga('send', 'pageview');
</script>


  </body>

</html>
